front/src/applications/operationalStudies/hooks/useScenarioData.ts

```diff
@@ -0,0 +1,207 @@
+      {
+        body: {
+          ids: timetable?.train_ids as number[],
+        },
+      },
+      {
+        skip: !timetable || !timetable.train_ids.length,
+      }
+    );
+
+  const { data: conflicts } = osrdEditoastApi.endpoints.getV2TimetableByIdConflicts.useQuery(
+    { id: scenario?.timetable_id as number, infraId: scenario?.infra_id as number },
+    { skip: !scenario }
+  );
+
+  const { data: projectionPath } = osrdEditoastApi.endpoints.getV2TrainScheduleByIdPath.useQuery(
+    {
+      id: trainIdUsedForProjection as number,
+      infraId: scenario?.infra_id as number,
+    },
+    {
+      skip: !scenario || !trainIdUsedForProjection,
+    }
+  );
+
+  const simulationResults = useSimulationResults();
+
+  const {
+    trainScheduleSummaries,
+    projectedTrains,
+    setTrainScheduleSummariesById,
+    setProjectedTrainsById,
+  } = useLazyLoadTrains({
+    infraId: scenario?.infra_id,
+    trainIdsToFetch,
+    setTrainIdsToFetch,
+    path: (projectionPath?.status === 'success' && projectionPath) || undefined,
+    trainSchedules,
+  });
+
+  useEffect(() => {
+    if (!timetable?.train_ids || !rawTrainSchedules) {
+      setTrainSchedules(undefined);
+    } else {
+      const sortedTrainSchedules = sortBy(rawTrainSchedules, 'start_time');
+      setTrainSchedules(sortedTrainSchedules);
+    }
+  }, [rawTrainSchedules]);
+
+  // first load of the trainScheduleSummaries
+  useEffect(() => {
+    if (trainSchedules && infra?.state === 'CACHED' && trainScheduleSummaries.length === 0) {
+      const trainIds = trainSchedules.map((trainSchedule) => trainSchedule.id);
+      setTrainIdsToFetch(trainIds);
+    }
+  }, [trainSchedules, infra?.state]);
+
+  useEffect(() => {
+    if (fetchTrainSchedulesError) {
+      dispatch(setFailure(castErrorToFailure(fetchTrainSchedulesError)));
+    }
+  }, [fetchTrainSchedulesError]);
+
+  useEffect(() => {
+    const removeTrains = (_trainIdsToRemove: number[]) => {
+      const trainSchedulesById = keyBy(trainSchedules, 'id');
+      const trainScheduleSummariesById = keyBy(trainScheduleSummaries, 'id');
+      const projectedTrainsById = keyBy(projectedTrains, 'id');
+
+      _trainIdsToRemove.forEach((trainId) => {
+        delete trainSchedulesById[trainId];
+        delete trainScheduleSummariesById[trainId];
+        delete projectedTrainsById[trainId];
+      });
+
+      const newTrainSchedules = Object.values(trainSchedulesById);
+      if (trainSchedules && newTrainSchedules.length !== trainSchedules.length)
+        setTrainSchedules(newTrainSchedules);
+
+      if (Object.values(trainScheduleSummariesById).length !== trainScheduleSummaries.length)
+        setTrainScheduleSummariesById(trainScheduleSummariesById);
+
+      if (Object.keys(projectedTrainsById).length !== projectedTrains.length)
+        setProjectedTrainsById(projectedTrainsById);
+
+      setTrainIdsToRemove([]);
+    };
+
+    if (trainIdsToRemove.length > 0) removeTrains(trainIdsToRemove);
```

Comment:
why not just an if statement with the body of the function  ?  why declare a function to call it once

Reply:
I thought it was more readable to have a function with a name instead of the code directly under the `if` line 134. I still prefer to have the function, what do you think ?

Reply:
same discussion as https://github.com/OpenRailAssociation/osrd/pull/8271#discussion_r1711663239